`TEMPLATE` function isn't working on certain `<<..>>` replacements

**Describe the bug**
A `TEMPLATE()` should be able to cover all cases where `<<..>>` was used before. The below example binding is erroring out,

```yaml
bindings:
  rw_key: <<keydist:Uniform(0,1000000000)->int>>; ToString() -> String
```
when modified as below,
```yaml
bindings:
  rw_key: TEMPLATE(keydist,Uniform(0,1000000000)->int); ToString() -> String
```
gives the following error when run with `-vv --show-stacktraces`:
<details>
    <summary>Expand/Collapse Stacktrace</summary>

    2788 DEBUG [scenarios:001] StrInterpolator removed template params after applying:{keycount=1000000000, valuecount=1000000000, keydist=Uniform(0,1000000000)->int, valdist=Uniform(0,1000000000)->int, database=baselines, table=keyvalue, read_ratio=5, keyspace=baselines, write_ratio=5}
    line 1:20 extraneous input '->' expecting {',', ')'}
    [->, int]
        2867 WARN  [scenarios:001] VirtDataDSL  Error while parsing flow:Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
        2878 ERROR [scenarios:001] SCENARIO     Error in scenario, shutting down. (javax.script.ScriptException: java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String')
        2878 DEBUG [scenarios:001] ScenarioController force stopping scenario scenario_20230207_091401_007
        2878 DEBUG [scenarios:001] ScenarioController Scenario force stopped.
        2879 DEBUG [scenarios:001] ScenarioController Requesting ScenarioController shutdown.
        2879 DEBUG [scenarios:001] SCENARIO     finishing scenario
        2879 INFO  [scenarios:001] SCENARIO     Scenario completed successfully, with 0 logical activities.
        2879 INFO  [scenarios:001] SCENARIO     scenario state: Errored
        2880 INFO  [scenarios:001] ANNOTATION   {"layer":"Scenario","start":1675779241971,"end":1675779243550,"labels":{"layer":"Scenario","state":"Errored","session":"scenario_20230207_091401_007","span":"interval","appname":"nosqlbench"},"details":{"command_line":"run\ndriver\u003djdbc\nworkload\u003d/Users/madhavan.sridharan/Data/Tools/sperf_and_nibbler/nosqlbench/local/jdbc/cockroachdb-keyvalue.yaml\ntags\u003dblock:main.*\nthreads\u003dAUTO\ncycles\u003d10\nurl\u003djdbc:postgresql://insectdb0-3153.g8z.cockroachlabs.cloud:26257/defaultdb?sslmode\u003dverify-ca\u0026password\u003dKw67BCIOCxLJrYLWtj2s4g\u0026user\u003dnewuser\ndatabaseName\u003ddefaultdb\nportNumber\u003d26257\nuser\u003dnewuser\npassword\u003dKw67BCIOCxLJrYLWtj2s4g\nsslmode\u003dverify-full\nserverName\u003dinsectdb\nsslrootcert\u003d/Users/madhavan.sridharan/Data/Tools/sperf_and_nibbler/nosqlbench/local/postgresql_certs/root.crt\n-vv\n--show-stacktraces"}}
        2880 DEBUG [scenarios:001] SCENARIO     removing scenario shutdown hook
        2880 DEBUG [scenarios:001] SCENARIO     ERRORED scenario run
        2886 DEBUG [scenarios:001] ExecutionResult populating ERROR scenario result
        2886 DEBUG [scenarios:001] ExecutionResult -- WARNING: Metrics which are taken per-interval (like histograms) will not have --
        2886 DEBUG [scenarios:001] ExecutionResult -- active data on this last report. (The workload has already stopped.) Record   --
        2886 DEBUG [scenarios:001] ExecutionResult -- metrics to an external format to see values for each reporting interval.      --
        2886 DEBUG [scenarios:001] ExecutionResult -- BEGIN METRICS DETAIL --
        2892 DEBUG [scenarios:001] ExecutionResult -- END METRICS DETAIL --
        2896 DEBUG [scenarios:001] SCENARIO     Summarizing counting metrics only to stdout:60 with scenario duration of 60000ms (<60000)
        2903 DEBUG [main] SCENARIOS    scenarios executor shutdown after 1602ms.
        2905 DEBUG [main] NBCLI        Total of 1 result object returned from ScenariosExecutor
        2906 INFO  [main] ScenariosResults results for scenario: name:'scenario_20230207_091401_007'
        2907 INFO  [main] ExecutionResult -- SCENARIO TOOK 1.579S --
        2911 INFO  [main] NBCLI        executions: 1 scenarios, 0 normal, 1 errored
        2912 WARN  [main] NBCLI        executions: 1 scenarios, 0 normal, 1 errored
        2916 ERROR [main] ERRORHANDLER Error from driver or included library: javax.script.ScriptException: java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
    java.lang.RuntimeException: javax.script.ScriptException: java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
        at io.nosqlbench.engine.core.lifecycle.scenario.Scenario.executeScenarioScripts(Scenario.java:335)
        at io.nosqlbench.engine.core.lifecycle.scenario.Scenario.runScenario(Scenario.java:269)
        at io.nosqlbench.engine.core.lifecycle.scenario.Scenario.call(Scenario.java:406)
        at io.nosqlbench.engine.core.lifecycle.scenario.Scenario.call(Scenario.java:63)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
    Caused by: javax.script.ScriptException: java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
        at io.nosqlbench.virtdata.core.bindings.VirtData.getOptionalMapper(VirtData.java:102)
        at io.nosqlbench.virtdata.core.templates.StringCompositor.lambda$new$1(StringCompositor.java:52)
        at java.base/java.util.HashMap.forEach(HashMap.java:1421)
        at io.nosqlbench.virtdata.core.templates.StringCompositor.<init>(StringCompositor.java:51)
        at io.nosqlbench.virtdata.core.templates.StringCompositor.<init>(StringCompositor.java:73)
        at io.nosqlbench.virtdata.core.templates.StringBindings.<init>(StringBindings.java:49)
        at io.nosqlbench.virtdata.core.templates.StringBindings.<init>(StringBindings.java:45)
        at io.nosqlbench.engine.api.templating.ParsedTemplateMap.lambda$applyTemplateFields$1(ParsedTemplateMap.java:124)
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
        at io.nosqlbench.engine.api.templating.ParsedTemplateMap.applyTemplateFields(ParsedTemplateMap.java:105)
        at io.nosqlbench.engine.api.templating.ParsedTemplateMap.<init>(ParsedTemplateMap.java:95)
        at io.nosqlbench.engine.api.templating.ParsedOp.<init>(ParsedOp.java:351)
        at io.nosqlbench.engine.api.activityimpl.uniform.StandardActivity.<init>(StandardActivity.java:81)
        at io.nosqlbench.engine.api.activityimpl.uniform.StandardActivityType.getActivity(StandardActivityType.java:59)
        at io.nosqlbench.engine.api.activityimpl.uniform.StandardActivityType.getActivity(StandardActivityType.java:36)
        at io.nosqlbench.engine.api.activityapi.core.ActivityType.getAssembledActivity(ActivityType.java:63)
        at io.nosqlbench.engine.core.lifecycle.activity.ActivityLoader.loadActivity(ActivityLoader.java:44)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.doStartActivity(ScenarioController.java:84)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.run(ScenarioController.java:137)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.run(ScenarioController.java:118)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.run(ScenarioController.java:147)
        at io.nosqlbench.engine.core.lifecycle.scenario.script.bindings.PolyglotScenarioController.run(PolyglotScenarioController.java:57)
        at <js>.:program(<eval>:1)
        at org.graalvm.polyglot.Context.eval(Context.java:399)
        at com.oracle.truffle.js.scriptengine.GraalJSScriptEngine.eval(GraalJSScriptEngine.java:478)
        at com.oracle.truffle.js.scriptengine.GraalJSScriptEngine.eval(GraalJSScriptEngine.java:420)
        at java.scripting/javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:247)
        at io.nosqlbench.engine.core.lifecycle.scenario.Scenario.executeScenarioScripts(Scenario.java:311)
        ... 7 more
    Caused by: java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
        at io.nosqlbench.virtdata.core.bindings.VirtData.getOptionalMapper(VirtData.java:102)
        at io.nosqlbench.virtdata.core.templates.StringCompositor.lambda$new$1(StringCompositor.java:52)
        at java.base/java.util.HashMap.forEach(HashMap.java:1421)
        at io.nosqlbench.virtdata.core.templates.StringCompositor.<init>(StringCompositor.java:51)
        at io.nosqlbench.virtdata.core.templates.StringCompositor.<init>(StringCompositor.java:73)
        at io.nosqlbench.virtdata.core.templates.StringBindings.<init>(StringBindings.java:49)
        at io.nosqlbench.virtdata.core.templates.StringBindings.<init>(StringBindings.java:45)
        at io.nosqlbench.engine.api.templating.ParsedTemplateMap.lambda$applyTemplateFields$1(ParsedTemplateMap.java:124)
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
        at io.nosqlbench.engine.api.templating.ParsedTemplateMap.applyTemplateFields(ParsedTemplateMap.java:105)
        at io.nosqlbench.engine.api.templating.ParsedTemplateMap.<init>(ParsedTemplateMap.java:95)
        at io.nosqlbench.engine.api.templating.ParsedOp.<init>(ParsedOp.java:351)
        at io.nosqlbench.engine.api.activityimpl.uniform.StandardActivity.<init>(StandardActivity.java:81)
        at io.nosqlbench.engine.api.activityimpl.uniform.StandardActivityType.getActivity(StandardActivityType.java:59)
        at io.nosqlbench.engine.api.activityimpl.uniform.StandardActivityType.getActivity(StandardActivityType.java:36)
        at io.nosqlbench.engine.api.activityapi.core.ActivityType.getAssembledActivity(ActivityType.java:63)
        at io.nosqlbench.engine.core.lifecycle.activity.ActivityLoader.loadActivity(ActivityLoader.java:44)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.doStartActivity(ScenarioController.java:84)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.run(ScenarioController.java:137)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.run(ScenarioController.java:118)
        at io.nosqlbench.engine.core.lifecycle.scenario.ScenarioController.run(ScenarioController.java:147)
        at io.nosqlbench.engine.core.lifecycle.scenario.script.bindings.PolyglotScenarioController.run(PolyglotScenarioController.java:57)
        at com.oracle.truffle.host.HostMethodDesc$SingleMethod$MHBase.invokeHandle(HostMethodDesc.java:350)
        at com.oracle.truffle.host.GuestToHostCodeCache$1.executeImpl(GuestToHostCodeCache.java:96)
        at com.oracle.truffle.host.GuestToHostRootNode.execute(GuestToHostRootNode.java:80)
        at com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:85)
        at com.oracle.truffle.api.impl.DefaultRuntimeAccessor$DefaultRuntimeSupport.callInlined(DefaultRuntimeAccessor.java:177)
        at com.oracle.truffle.host.GuestToHostRootNode.guestToHostCall(GuestToHostRootNode.java:102)
        at com.oracle.truffle.host.HostMethodDesc$SingleMethod$MHBase.invokeGuestToHost(HostMethodDesc.java:386)
        at com.oracle.truffle.host.HostExecuteNode.doInvoke(HostExecuteNode.java:871)
        at com.oracle.truffle.host.HostExecuteNode.doOverloadedCached(HostExecuteNode.java:288)
        at com.oracle.truffle.host.HostExecuteNodeGen.executeAndSpecialize(HostExecuteNodeGen.java:241)
        at com.oracle.truffle.host.HostExecuteNodeGen.execute(HostExecuteNodeGen.java:97)
        at com.oracle.truffle.host.HostObject.invokeMember(HostObject.java:454)
        at com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached.invokeMemberNode_AndSpecialize(HostObjectGen.java:2696)
        at com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached.invokeMember(HostObjectGen.java:2678)
        at com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch.invokeMember(InteropLibraryGen.java:8328)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode$ForeignInvokeNode.executeCall(JSFunctionCallNode.java:1534)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode.executeAndSpecialize(JSFunctionCallNode.java:306)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode.executeCall(JSFunctionCallNode.java:251)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode$InvokeNode.execute(JSFunctionCallNode.java:740)
        at com.oracle.truffle.js.nodes.access.JSWriteCurrentFrameSlotNodeGen.execute_generic4(JSWriteCurrentFrameSlotNodeGen.java:124)
        at com.oracle.truffle.js.nodes.access.JSWriteCurrentFrameSlotNodeGen.execute(JSWriteCurrentFrameSlotNodeGen.java:43)
        at com.oracle.truffle.js.nodes.access.JSWriteCurrentFrameSlotNodeGen.executeVoid(JSWriteCurrentFrameSlotNodeGen.java:330)
        at com.oracle.truffle.js.nodes.binary.DualNode.execute(DualNode.java:118)
        at com.oracle.truffle.js.nodes.function.FunctionBodyNode.execute(FunctionBodyNode.java:73)
        at com.oracle.truffle.js.nodes.function.FunctionRootNode.executeInRealm(FunctionRootNode.java:149)
        at com.oracle.truffle.js.runtime.JavaScriptRealmBoundaryRootNode.execute(JavaScriptRealmBoundaryRootNode.java:88)
        at com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:85)
        at com.oracle.truffle.api.impl.DefaultDirectCallNode.call(DefaultDirectCallNode.java:59)
        at com.oracle.truffle.js.lang.JavaScriptLanguage$ParsedProgramRoot.execute(JavaScriptLanguage.java:245)
        at com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:85)
        at com.oracle.truffle.api.impl.DefaultCallTarget.call(DefaultCallTarget.java:102)
        at com.oracle.truffle.polyglot.PolyglotContextImpl.eval(PolyglotContextImpl.java:1470)
        at com.oracle.truffle.polyglot.PolyglotContextDispatch.eval(PolyglotContextDispatch.java:63)
        ... 12 more
        2930 ERROR [main] ERRORHANDLER cause (see stack trace for details):java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
    javax.script.ScriptException: java.lang.RuntimeException: Error while parsing binding specification 'Uniform(0,1000000000->int); ToString() -> String': java.lang.RuntimeException: Error parsing input 'Uniform(0,1000000000->int); ToString() -> String'
    Process finished with exit code 2
</details>

**To Reproduce**
Steps to reproduce the behavior:
See above bug description to repro the problem behavior.

**What was Expected**
The above form should transparently work.

**Additional context**
Add any other context about the problem here.
OS: macOS Monteray
environment: n/a
version info (`./nb --version` or `java -jar nb.jar --version`): `5.17.1-SNAPSHOT`

**Screenshots, if applicable**
n/a

--- a/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/activityconfig/OpsLoader.java
+++ b/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/activityconfig/OpsLoader.java
@@ -20,6 +20,7 @@ import com.amazonaws.util.StringInputStream;
 import com.google.gson.GsonBuilder;
 import io.nosqlbench.nb.api.nbio.Content;
 import io.nosqlbench.nb.api.nbio.NBIO;
+import io.nosqlbench.nb.api.advisor.NBAdvisorOutput;
 import io.nosqlbench.nb.api.advisor.NBAdvisorException;
 import io.nosqlbench.nb.api.errors.BasicError;
 import io.nosqlbench.adapters.api.activityconfig.rawyaml.RawOpsDocList;
@@ -29,6 +30,7 @@ import io.nosqlbench.adapters.api.activityconfig.yaml.OpsDocList;
 import io.nosqlbench.adapters.api.templating.StrInterpolator;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Level;
 import org.snakeyaml.engine.v2.api.Load;
 import org.snakeyaml.engine.v2.api.LoadSettings;
 import scala.Option;
@@ -69,6 +71,9 @@ public class OpsLoader {
         logger.trace(() -> "Applying string transformer to data:" + sourceData);
         StrInterpolator transformer = new StrInterpolator(params);
         String data = transformer.apply(sourceData);
+        NBAdvisorOutput.render(Level.INFO,"Transform:");
+        NBAdvisorOutput.render(Level.INFO,"From: "+sourceData);
+        NBAdvisorOutput.render(Level.INFO,"  To: "+data);
         if (srcuri!=null) {
             logger.info("workload URI: '" + srcuri + "'");
         }

--- a/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/templating/StrInterpolator.java
+++ b/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/templating/StrInterpolator.java
@@ -17,10 +17,12 @@
 package io.nosqlbench.adapters.api.templating;
 
 import io.nosqlbench.nb.api.engine.activityimpl.ActivityDef;
+import io.nosqlbench.nb.api.advisor.NBAdvisorOutput;
 import org.apache.commons.text.StrLookup;
 import org.apache.commons.text.StringSubstitutor;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Level;
 
 import java.util.*;
 import java.util.function.Function;
@@ -31,12 +33,6 @@ public class StrInterpolator implements Function<String, String> {
     private final MultiMap multimap = new MultiMap();
     private final StringSubstitutor substitutor =
         new StringSubstitutor(multimap, "<<", ">>", '\\')
-            .setEnableSubstitutionInVariables(true)
-            .setEnableUndefinedVariableException(true)
-            .setDisableSubstitutionInValues(true);
-
-    private final StringSubstitutor substitutor2 =
-        new StringSubstitutor(multimap, "TEMPLATE(", ")", '\\')
             .setEnableSubstitutionInVariables(true)
             .setEnableUndefinedVariableException(true)
             .setDisableSubstitutionInValues(true);
@@ -58,10 +54,23 @@ public class StrInterpolator implements Function<String, String> {
 
     @Override
     public String apply(String raw) {
-        String after = substitutor.replace(substitutor2.replace(raw));
+        String after = matchTemplates(raw);
         while (!after.equals(raw)) {
             raw = after;
-            after = substitutor.replace(substitutor2.replace(raw));
+            after = matchTemplates(raw);
+        }
+	raw = after;
+	String original = raw;
+        after = substitutor.replace(raw);
+        while (!after.equals(raw)) {
+            raw = after;
+            after = substitutor.replace(raw);
+        }
+        if ( !original.equals(after)) {
+            NBAdvisorOutput.output(Level.WARN,"Transform <<key:value>>");
+            NBAdvisorOutput.output(Level.WARN,"From: "+original);
+            NBAdvisorOutput.output(Level.WARN,"  To: "+after);
+            NBAdvisorOutput.test("Using the deprecated template for of <<key:value>> please use TEMPLATE(key,value)");
         }
         return after;
     }
@@ -74,6 +83,41 @@ public class StrInterpolator implements Function<String, String> {
         LinkedHashMap<String, String> details = new LinkedHashMap<>();
 
         return details;
+    }
+
+    public String matchTemplates(String original) {
+        String line = original;
+        int length = line.length();
+        int i = 0;
+        while (i < length) {
+            // Detect an instance of "TEMPLATE("
+            if (line.startsWith("TEMPLATE(", i)) {
+                int start = i + "TEMPLATE(".length();
+                int openParensCount = 1; // We found one '(' with "TEMPLATE("
+                // Find the corresponding closing ')' for this TEMPLATE instance
+                int j = start;
+                while (j < length && openParensCount > 0) {
+                    if (line.charAt(j) == '(') {
+                        openParensCount++;
+                    } else if (line.charAt(j) == ')') {
+                        openParensCount--;
+                    }
+                    j++;
+                }
+                // `j` now points just after the closing ')' of this TEMPLATE
+                if (openParensCount == 0) {
+                    String templateContent = line.substring(start, j - 1);
+                    // Resolve the template content
+                    String resolvedContent = multimap.lookup(templateContent);
+                    line = line.substring(0, i) + resolvedContent + line.substring(j);
+                    // Update `length` and `i` based on the modified `line`
+                    i += resolvedContent.length() - 1;
+                    length = line.length();
+                }
+            }
+            i++;
+        }
+        return line;
     }
 
     public static class MultiMap extends StrLookup<String> {
@@ -131,6 +175,5 @@ public class StrInterpolator implements Function<String, String> {
 
         }
     }
-
 
 }

--- a/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/templating/StrInterpolator.java
+++ b/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/templating/StrInterpolator.java
@@ -59,8 +59,8 @@ public class StrInterpolator implements Function<String, String> {
             raw = after;
             after = matchTemplates(raw);
         }
-	raw = after;
-	String original = raw;
+        raw = after;
+        String original = raw;
         after = substitutor.replace(raw);
         while (!after.equals(raw)) {
             raw = after;

--- a/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/templating/StrInterpolator.java
+++ b/nb-apis/adapters-api/src/main/java/io/nosqlbench/adapters/api/templating/StrInterpolator.java
@@ -96,13 +96,22 @@ public class StrInterpolator implements Function<String, String> {
                 int openParensCount = 1; // We found one '(' with "TEMPLATE("
                 // Find the corresponding closing ')' for this TEMPLATE instance
                 int j = start;
+                int k = start;
                 while (j < length && openParensCount > 0) {
                     if (line.charAt(j) == '(') {
                         openParensCount++;
                     } else if (line.charAt(j) == ')') {
+                        k = j;
                         openParensCount--;
                     }
                     j++;
+                }
+                // check for case of not enough ')'
+                if (openParensCount > 0 ) {
+                    if ( k != start ) {
+                        j = k + 1;
+                    }
+                    openParensCount = 0;
                 }
                 // `j` now points just after the closing ')' of this TEMPLATE
                 if (openParensCount == 0) {
@@ -111,7 +120,7 @@ public class StrInterpolator implements Function<String, String> {
                     String resolvedContent = multimap.lookup(templateContent);
                     line = line.substring(0, i) + resolvedContent + line.substring(j);
                     // Update `length` and `i` based on the modified `line`
-                    i += resolvedContent.length() - 1;
+                    // i += resolvedContent.length() - 1;
                     length = line.length();
                 }
             }

--- a/nb-apis/adapters-api/src/test/java/io/nosqlbench/adapters/api/templating/StrInterpolatorTest.java
+++ b/nb-apis/adapters-api/src/test/java/io/nosqlbench/adapters/api/templating/StrInterpolatorTest.java
@@ -78,6 +78,13 @@ public class StrInterpolatorTest {
     }
 
     @Test
+    public void shouldMatchNestedParens() {
+        StrInterpolator interp = new StrInterpolator(abcd);
+        String a = interp.apply("TEMPLATE(keydist,Uniform(0,1000000000)->int);");
+        assertThat(a).isEqualTo("Uniform(0,1000000000)->int;");
+    }
+
+    @Test
     public void shouldReturnWarningWhenUnmatched() {
         StrInterpolator interp = new StrInterpolator(abcd);
         String a = interp.apply("<<nokeymatchesthis>>");

--- a/nb-apis/adapters-api/src/test/java/io/nosqlbench/adapters/api/templating/StrInterpolatorTest.java
+++ b/nb-apis/adapters-api/src/test/java/io/nosqlbench/adapters/api/templating/StrInterpolatorTest.java
@@ -19,6 +19,7 @@ package io.nosqlbench.adapters.api.templating;
 import org.junit.jupiter.api.Test;
 
 import java.util.ArrayList;
+import java.util.LinkedHashMap;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -136,20 +137,22 @@ public class StrInterpolatorTest {
         assertThat(a).isEqualTo("'Key': 'Value'.'Stuff'");
     }
 
-//    @Test
-//    public void shouldExpandNestedTemplates() {
-//        String a = interp.apply("-TEMPLATE(akey,TEMPLATE(dkey,whee)-");
-//        assertThat(a).isEqualTo("-aval1-");
-//        String b = interp.apply("-TEMPLATE(unknown,TEMPLATE(bkey,whee))-");
-//        assertThat(b).isEqualTo("-bval1-");
-//    }
-//
+    @Test
+    public void shouldExpandNestedTemplates() {
+        StrInterpolator interp = new StrInterpolator(abcd);
+        String a = interp.apply("-TEMPLATE(akey,TEMPLATE(dkey,whee)-");
+        assertThat(a).isEqualTo("-aval1-");
+        String b = interp.apply("-TEMPLATE(unknown,TEMPLATE(bkey,whee))-");
+        assertThat(b).isEqualTo("-bval1-");
+    }
+
 //    @Test
 //    public void shouldGetBasicDetails() {
+//        StrInterpolator interp = new StrInterpolator(abcd);
 //        LinkedHashMap<String, String> details = interp.getTemplateDetails("-TEMPLATE(akey,TEMPLATE(dkey,whee)-");
 //        assertThat(details).containsOnlyKeys("akey","dkey");
 //        assertThat(details).containsValues("test1");
 //
 //    }
-//
+
 }

--- a/nb-apis/nb-api/src/main/java/io/nosqlbench/nb/api/advisor/NBAdvisorOutput.java
+++ b/nb-apis/nb-api/src/main/java/io/nosqlbench/nb/api/advisor/NBAdvisorOutput.java
@@ -1,0 +1,50 @@
+package io.nosqlbench.nb.api.advisor;
+
+/*
+ * Copyright (c) nosqlbench
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Level;
+
+public class NBAdvisorOutput {
+    private final static Logger logger = LogManager.getLogger("ADVISOR");
+
+    public static void render(Level level,String message) {
+        if (NBAdvisorLevel.get() == NBAdvisorLevel.validate) {
+            output(level, message);
+        }
+    }
+    
+    public static void test(String message) {
+	output(Level.WARN, message);
+        if (NBAdvisorLevel.get() == NBAdvisorLevel.enforce) {
+	    throw new NBAdvisorException(message, 2);
+        }
+    }
+    
+    public static void output(Level level,String message) {
+	if (level ==  Level.INFO) {
+	    logger.info(message);
+	} else if (level == Level.WARN) {
+	    logger.warn(message);
+	} else if (level == Level.ERROR) {
+	    logger.error(message);
+}
+    }
+
+}

--- a/nb-engine/nb-engine-cli/src/test/resources/activities/example_scenarios_templatevars.yaml
+++ b/nb-engine/nb-engine-cli/src/test/resources/activities/example_scenarios_templatevars.yaml
@@ -6,5 +6,6 @@ scenarios:
 bindings:
   cycle: Identity()
   name: NumberNameToCycle()
+  rw_key: TEMPLATE(keydist,Uniform(0,1000000000)->int); ToString() -> String
 ops:
   cycle: "cycle {cycle} TEMPLATE(tvar1,def1) TEMPLATE(tvar1)\n"

--- a/nb-engine/nb-engine-core/src/main/java/io/nosqlbench/engine/api/scenarios/NBCLIScenarioPreprocessor.java
+++ b/nb-engine/nb-engine-core/src/main/java/io/nosqlbench/engine/api/scenarios/NBCLIScenarioPreprocessor.java
@@ -312,10 +312,6 @@ public class NBCLIScenarioPreprocessor {
         return parsedStep;
     }
 
-    private static final Pattern templatePattern = Pattern.compile("TEMPLATE\\((.+?)\\)");
-    private static final Pattern innerTemplatePattern = Pattern.compile("TEMPLATE\\((.+?)\\)");
-    private static final Pattern templatePattern2 = Pattern.compile("<<(.+?)>>");
-
     public static List<WorkloadDesc> filterForScenarios(List<Content<?>> candidates) {
 
         List<Path> yamlPathList = candidates.stream().map(Content::asPath).toList();
@@ -437,28 +433,43 @@ public class NBCLIScenarioPreprocessor {
 
     }
 
+    //    private static final Pattern templatePattern = Pattern.compile("TEMPLATE\\((.+?)\\)");
+    //    private static final Pattern innerTemplatePattern = Pattern.compile("TEMPLATE\\((.+?)\\)");
+    private static final Pattern templatePattern2 = Pattern.compile("<<(.+?)>>");
+
     public static Map<String, String> matchTemplates(String line, Map<String, String> templates) {
-        Matcher matcher = templatePattern.matcher(line);
-
-        while (matcher.find()) {
-            String match = matcher.group(1);
-
-            Matcher innerMatcher = innerTemplatePattern.matcher(match);
-            if (innerMatcher.find()) {
-                String innerMatch = innerMatcher.group(1);
-                templates = matchTemplates("TEMPLATE(" + innerMatch + ")", templates);
-                String resolvedInner = templates.getOrDefault(innerMatch.split("[,:]")[0], "");
-                match = match.replace("TEMPLATE(" + innerMatch + ")", resolvedInner);
+        int length = line.length();
+        int i = 0;
+        while (i < length) {
+            // Detect an instance of "TEMPLATE("
+            if (line.startsWith("TEMPLATE(", i)) {
+                int start = i + "TEMPLATE(".length();
+                int openParensCount = 1; // We found one '(' with "TEMPLATE("
+                // Find the corresponding closing ')' for this TEMPLATE instance
+                int j = start;
+                while (j < length && openParensCount > 0) {
+                    if (line.charAt(j) == '(') {
+                        openParensCount++;
+                    } else if (line.charAt(j) == ')') {
+                        openParensCount--;
+                    }
+                    j++;
+                }
+                // `j` now points just after the closing ')' of this TEMPLATE
+                if (openParensCount == 0) {
+                    String templateContent = line.substring(start, j - 1);
+                    // Resolve the template content
+                    String resolvedContent = resolveTemplate(templateContent, templates);
+                    line = line.substring(0, i) + resolvedContent + line.substring(j);
+                    // Update `length` and `i` based on the modified `line`
+                    i += resolvedContent.length() - 1;
+                    length = line.length();
+                }
             }
-
-            String[] matchArray = match.split("[,:]");
-            if (matchArray.length == 1) {
-                matchArray = new String[]{matchArray[0], ""};
-            }
-            templates.put(matchArray[0], matchArray[1]);
+            i++;
         }
-        matcher = templatePattern2.matcher(line);
 
+        Matcher matcher = templatePattern2.matcher(line);
         while (matcher.find()) {
             String match = matcher.group(1);
             String[] matchArray = match.split(":");
@@ -471,5 +482,17 @@ public class NBCLIScenarioPreprocessor {
         return templates;
     }
 
+    private static String resolveTemplate(String content, Map<String, String> templates) {
+        String[] parts = content.split("[,:]", 2);
+        String key = parts[0];
+        String value = parts.length > 1 ? parts[1] : "";
+
+        // Store or retrieve the resolved value for the template key
+        if (!templates.containsKey(key)) {
+            templates.put(key, value);
+        }
+        
+        return templates.getOrDefault(key, "");
+    }
 
 }

